### DISCUSSION TITLE
potential libraries to include

### DISCUSSION THREAD

**Discussion by dcnieho** [pthom/imgui_bundle#84] [Ideas]
Hi Pascal,

I saw https://github.com/CoolLibs/imgui_gradient, which might be of interest to include
And it could be worth considering to replace https://github.com/dfranx/ImFileDialog with https://github.com/time-killer-games/libfiledialogs

Cheers!

**Comment by pthom** [comment]
Hi Dee,

I'm gonna transform this issue into a discussion inside discussions/ideas, as it will better fit as a discussion.

Now, concerning those two:

* https://github.com/CoolLibs/imgui_gradient
I'll keep this in the idea box. The API is quite extensive, and thus not immediate to bind. Will see later if more users ask for it.

* And it could be worth considering to replace https://github.com/dfranx/ImFileDialog with https://github.com/time-killer-games/libfiledialogs

A good idea. However, I recommend you take a look at the bindings for [Portable File Dialogs](https://github.com/pthom/imgui_bundle/blob/main/bindings/imgui_bundle/portable_file_dialogs.pyi). They are easy to use, and use the OS dialogs, which is what the final users want, IMHO.

**Reply by dcnieho** [reply]
Yeah, I agree that Portable File Dialogs is likely what many users want. But since you offer ImFileDialog already and that seems unmaintained and somewhat buggy, swapping it out for the other one could be useful.

One problem i have had is making a selection dialogue for multiple folders--that seems to not be a native thing so portable file dialogues doesn't do it. I'm not sure whether libfiledialogs does it, never tried, and have my own solution. But at least I'd be implementable there if i  ever want to.

**Comment by dcnieho** [comment]
Here's another one you could consider replacing, the colorized text editor.
As per https://github.com/BalazsJako/ImGuiColorTextEdit/issues/128, it has been unmaintained for years and this won't change.
dfranx has a form that has advanced a lot, but also seems dead for the last two years: https://github.com/dfranx/ImGuiColorTextEdit
This one however seems to be actively worked on: https://github.com/santaclose/ImGuiColorTextEdit

Since the current editor has some annoying bugs (like that you can't mouse-click to the end of a line), it is worth considering to replace with santaclose's fork.

**Reply by pthom** [reply]
ImGuiColorTextEdit is now based on the santaclose fork

**Reply by dcnieho** [reply]
Super! Do you plan to do a new release soon so i can try it out?

**Reply by dcnieho** [reply]
And i see you just did last week :p I shall try it out!

**Comment by pthom** [comment]
Hi Dee,

Concerning ImGuiColorTextEdit, I know that it is not maintained. I already added a list of [interesting forks](https://github.com/pthom/ImGuiColorTextEdit/blob/imgui_bundle/Forks.md), and santaclose is part of them. However, what this fork changes is not well documented (apart from reading the commit log). As written in the forks.md file I suspect it mainly adds support for multicursor, auto-indentation and python.

I think I took inspiration from for the python language support which I added in my fork.

Concerning the mouse click at end of line, I fixed it about a month ago in https://github.com/pthom/ImGuiColorTextEdit/commit/0708c7acadb9ce4557374bff56b8b915b0ab8487

I also improved the default color palette in  https://github.com/pthom/ImGuiColorTextEdit/commit/bc44d0ac7087b0d989565156ccb4fa66128a6b11

---

Concerning ImFileDialog's fork https://github.com/time-killer-games/libfiledialogs, its code is too bloated (it mixes the code from SDL, imgui, lunasvg, as well as some dll all into the same repository), I cannot use it as is. Its author [contacted me](https://github.com/dfranx/ImFileDialog/issues/32), and I wrote him about this.

**Reply by dcnieho** [reply]
Ah ok! Didn't know you fixed that bug (just tried the demo and indeed) and indeed, the situation with not well-documented and diverging forks is not ideal.
Too bad that ImFileDialog fork isn't integratable, but i understand.

**Comment by district10** [comment]
Is there a guide about how to "bind some c++ lib to python, then use it alongside imgui_bundle"? (or is it possible for imgui_bundle to leave some area for other program to draw on)

Like integrating <https://github.com/glampert/debug-draw> into imgui, something like this: <https://www.youtube.com/watch?v=aE8p9937syw>

**Reply by district10** [reply]
Surfing GitHub for a while (months..) and I found [vedo](https://vedo.embl.es/#gallery) is what i'm looking for. (there's also an imgui integration via polyscope, example [here](https://github.com/marcomusy/vedo/blob/master/examples/other/vpolyscope.py)). Hope one day imgui-bundle can be integrated with vedo. 😄

**Reply by pthom** [reply]
> "bind some c++ lib to python, then use it alongside imgui_bundle"

See doc [about litgen](https://pthom.github.io/litgen/litgen_book/00_00_intro.html), the generator behind ImGui Bundle 

> or is it possible for imgui_bundle to leave some area for other program to draw on

https://traineq.org/ImGuiBundle/emscripten/bin/demo_custom_background.html

https://github.com/pthom/imgui_bundle/blob/main/bindings/imgui_bundle/demos_cpp/demos_immapp/demo_custom_background.cpp

**Reply by dcnieho** [reply]
https://github.com/marcomusy/vedo/issues/933

**Reply by pthom** [reply]
@dcnieho : I just added support for MatPlotLib images:
https://github.com/pthom/imgui_bundle/commit/e2dab06f802df3afbece119902755c8ca415949b

If you are interested in vedo, you could take inspiration from this

**Reply by pthom** [reply]
> "bind some c++ lib to python, then use it alongside imgui_bundle"
Also see [developer doc](https://pthom.github.io/imgui_bundle/devel_docs/bindings.html#bindings_intro)

**Comment by district10** [comment]
I just took a look at recent imgui_bundle, the immvision is awesome! I use some similiar panzoom viewer (in browser) to inspect SVG, really helpful when a closer look is needed.

Then, how about [nanovg](https://github.com/memononen/nanovg). 😅

**Reply by pthom** [reply]
I added NanoVG support recently. It was quite a complex and long task, but it works.

see [python bindings](https://github.com/pthom/imgui_bundle/blob/main/bindings/imgui_bundle/nanovg.pyi), code of [python demos](https://github.com/pthom/imgui_bundle/tree/main/bindings/imgui_bundle/demos_python/demos_nanovg), code of [C++ demos](https://github.com/pthom/imgui_bundle/tree/main/bindings/imgui_bundle/demos_cpp/demos_nanovg), online [full demo](https://traineq.org/ImGuiBundle/emscripten/bin/demo_nanovg_full.html), online [simple demo](https://traineq.org/ImGuiBundle/emscripten/bin/demo_nanovg_heart.html), and [API for integration with ImGui](https://github.com/pthom/imgui_bundle/blob/main/external/nanovg/nvg_imgui/nvg_imgui.h). Works on Linux, Windows, macOS, emscripten, iOS and Android (OpenGL only)

**Reply by pthom** [reply]
Please, if possible, post images of your results - with NanoVG or other - in the [gallery](https://github.com/pthom/imgui_bundle/discussions/107)

**Reply by dcnieho** [reply]
On windows, i have a hiDPI problem with the full demo. Well, parts of it, the widgets and text are way too small.

**Reply by pthom** [reply]
> On windows, i have a hiDPI problem with the full demo. Well, parts of it, the widgets and text are way too small.

This was an issue with the official demo of NanoVG, which does not support HighDPI. I did a recent adaptation

**Comment by pthom** [comment]
@district10 many thanks for those nice suggestions. I’m on holiday so it might take me a while to study those, but I will have a look.

**Comment by tingspain** [comment]
Hello everyone,

I’d like to ask if there is any possibility of adding the Imoguizmo library to the ImGui Bundle.

Imoguizmo is a view gizmo manipulator similar to the one used in Blender, which could be a valuable addition for users working with 3D transformations in ImGui.

Here’s the repository for reference:
🔗 [Imoguizmo GitHub](https://github.com/fknfilewalker/imoguizmo)

Would this be something worth considering for integration? I’d love to hear your thoughts.

Thanks for your time!

**Comment by pthom** [comment]
Hi @tingspain
> I’d like to ask if there is any possibility of adding the Imoguizmo library to the ImGui Bundle.

ImGui Bundle already includes [ImGuizmo](https://github.com/CedricGuillemet/ImGuizmo). It is more well known than ImoGuizmo, and provides a Guizmo. Did you try it?

**Reply by tingspain** [reply]
Hi @pthom ,

Sorry for my late reply! I really appreciate your response.

Yes, I’m aware that **ImGuizmo** is already included in the **ImGui Bundle**, and I appreciate its integration. I’m currently using **ImGuizmo** (attached a screenshot for reference), and it has been great. ^__^

However, I wanted to suggest **ImoGuizmo** as an additional option because it is a more modern alternative and offers greater flexibility in certain cases. My goal isn’t to replace **ImGuizmo** but rather to expand the available tools within **ImGui Bundle**, giving users more choices—especially those who might prefer its different approach to gizmo look&feel.

Would it be possible to consider adding it alongside **ImGuizmo**?

Looking forward to your thoughts! 

Thanks again!



![imguizmo_low](https://github.com/user-attachments/assets/0de80925-8854-42ea-9931-b18300fec8ec)

**Comment by pthom** [comment]
@tingspain :

I tried ImOGuizmo, and I feel like I lost my time. Did you try it on your side?

The Readme suggests that it handle *mouse drag* interactions, however it seems like it does not.

See https://github.com/fknfilewalker/imoguizmo/issues/2

**Reply by pthom** [reply]
If you need to use it in C++, you might directly include it in your project (it is header only).
If you want to use it in Python, I would says that porting this library to python (its code is only 300 lines long) might be a better approach; and I suspect that a LLM would output a good first conversion.

**Reply by tingspain** [reply]
Hi @pthom ,

First of all, I’m really sorry if I made you feel like you wasted your time—that was never my intention. I truly appreciate the time and effort you put into testing ImoGuizmo and sharing your findings.

You’re absolutely right—I’m writing my project in Python, and I had the same thought that porting ImoGuizmo from scratch might be the better approach. In fact, I went ahead and I started implementing it it in Python a few days ago! When I have a clean version of the code, I’ll share it with the ImGui Bundle community.

Thanks again for your insights and feedback—I really appreciate it!

![pyImoguizmo_1024](https://github.com/user-attachments/assets/452ae99d-5f5b-4279-a5cf-a31723e3a52f)

**Reply by tingspain** [reply]
Hi everyone! 

I’m excited to share that I have partially ported ImOGuizmo to Python and recently released it on GitHub as PyImoGuizmo! 🚀

**🔹 What is PyImoGuizmo?**
PyImoGuizmo is a Python library that brings the interactive orientation gizmo from ImOGuizmo to Python applications using ImGui. It is built on PyGLM for efficient matrix operations and is designed to be simple and easy to integrate into ImGui-based projects.

**🔹 What’s Included?**
Alongside the library, I’ve also provided an example app that demonstrates:
- How to use PyImoGuizmo for camera control in a 3D viewport
- Integration with imgui_bundle for the UI
- Use of ModernGL for rendering

**🔹 Where to Find It?**
Check out the project here:
🔗 [GitHub Repository – PyImoGuizmo](https://github.com/tingspain/PyImOGuizmo)

If you're interested, feel free to try it out! Any feedback, contributions, or suggestions are more than welcome. 😊

Looking forward to hearing your thoughts!

**Comment by samuelvenable** [comment]
Hi, I wanted to bring to everyone's attention that https://github.com/time-killer-games/libfiledialogs is no longer available. The URL is dead because my old time-killer-games account has been deleted. When my email got hacked it caused me to overreact and delete all my online accounts. I eventually got over it and signed up to GitHub again.

The new repository lives here:

https://github.com/samuelvenable/SDL2-ImGui-FileDialogs

...or if you want an SDL3 verson:

https://github.com/samuelvenable/SDL3-ImGui-FileDialogs

The two are identical and equally maintained. Only difference is the SDL version.

@pthom wanted your thoughts on whether you decision was final about not including this in the repos.

Edit:

Additionally, I wanted feedback on whether I should add back a means to use an OS-native title bar instead of just the ImGui-based one. The ImGui-based title bar doesn't allow dragging the window with a mouse down event, so the window can't be moved in its current state, unless you set the `IMGUI_DIALOG_PARENT` `std::uintptr_t` environment variable, which allows you to move the dialog by the specified parent window's title bar, keeping the dialog in the center position of the parent. Of course, this only works on macOS and X11 platforms, because on Windows the parent window is disabled from moving as well whenever a dialog is opened. Standard Windows dialog parent behavior.

**Comment by pthom** [comment]
Hi @samuelvenable 

There are already two file dialogs libraries in imgui bundle, which is enough IMHO. Furthermore, your project files & folder organization makes it hard to include (too many included third parties / fonts / lib binaries). So yes, my decision is final, I'm sorry.

**Reply by samuelvenable** [reply]
All good!

**Comment by claydugo** [comment]
[ImAnim](https://github.com/soufianekhiat/ImAnim) came across my feed this morning and it seemed cool and worth adding to the discussion page here.
